### PR TITLE
feat(payment): PI-517 Move Stripe UPE customer and payment strategy t…

### PR DESCRIPTION
**Pull request by bc-sebastianszafraniec**
…o package

There are also included changes from 4 PR’s related to StripeUPE, which were merged during work on this PR:
PI-758 - https://github.com/bigcommerce/checkout-sdk-js/pull/2201/
PI-889 - https://github.com/bigcommerce/checkout-sdk-js/pull/2213
PI-976 - https://github.com/bigcommerce/checkout-sdk-js/pull/2236
PI-924 - https://github.com/bigcommerce/checkout-sdk-js/pull/2242

Checkout-js part: https://github.com/bigcommerce/checkout-js/pull/1585

Related PR:  https://github.com/bigcommerce/checkout-sdk-js/pull/2330

## What?
Create StripevUPE package in checkout-sdk-js and move to it StripeUPE payment and customer strategies.

## Why?
For better separation between payment integrations and core Checkout SDK

## Testing / Proof
Manually tested

@bigcommerce/team-checkout @bigcommerce/team-payments

### CODE REVIEW

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 1]
```diff
@@ -87,20 +84,21 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
             stripeupe.onError?.(error),
         );
 
-        this._unsubscribe = await this._store.subscribe(
-            async (_state) => {
+        this._unsubscribe = this.paymentIntegrationService.subscribe(
+            // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:88]: You can disable this rule in the appropriate `.eslintrs.json` file

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 2]
```diff
@@ -235,27 +227,25 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
         }
     }
 
-    private async _executeWithAPM(methodId: string): Promise<InternalCheckoutSelectors> {
-        const state = this._store.getState();
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+    private async _executeWithAPM(methodId: string): Promise<any> {
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:230]: Can we use some `interface` instead of `any` type here?

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 3]
```diff
@@ -338,15 +324,17 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
...
-                params: { method: methodId },
-            }),
-        );
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+        const state = await this.paymentIntegrationService.loadPaymentMethod(gatewayId, {
+            params: { method: methodId },
+        });
+
+        const paymentMethod = this.paymentIntegrationService
+            .getState()
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:332]: You already have `state` above, try to use it

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 4]
```diff
@@ -338,15 +324,17 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
...
+            .getPaymentMethodOrThrow(methodId);
+
         const {
             initializationData: { stripePublishableKey, stripeConnectedAccount, shopperLanguage },
-        } = paymentMethod;
+        } = paymentMethod as StripeUPEPaymentMethod;
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:337]: Can we avoid using `as` here?

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 5]
```diff
@@ -422,13 +407,15 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
...
+    // TODO: complexity of _processAdditionalAction method
+    // eslint-disable-next-line complexity
     private async _processAdditionalAction(
         error: Error,
         methodId: string,
         shouldSaveInstrument = false,
         shouldSetAsDefaultInstrument = false,
-    ): Promise<InternalCheckoutSelectors | never> {
-        if (!(error instanceof RequestError)) {
+    ): Promise<any> {
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:417]: The same

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 6]
```diff
@@ -517,8 +502,8 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
         error: Error,
         methodId?: string,
         shouldSetAsDefaultInstrument = false,
-    ): Promise<InternalCheckoutSelectors | never> {
-        if (!(error instanceof RequestError)) {
+    ): Promise<any> {
```
Comment (bc-nick) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:505]: I see this method can have another returning type

File: docs/interfaces/Customer.md [hunk 7]
```diff
@@ -13,7 +13,6 @@
 - [fullName](Customer.md#fullname)
 - [id](Customer.md#id)
 - [isGuest](Customer.md#isguest)
-- [isStripeLinkAuthenticated](Customer.md#isstripelinkauthenticated)
```
Comment (bc-peng) [inline docs/interfaces/Customer.md:16]: This file is managed automatically.
We normally don't touch it in a PR.
Is this change intentional?

**Comment by bc-peng** [comment]
Also, could you please fill in Testing / Proof section?
If no manual testing happened, you could simply put `CI checks` there.

File: docs/interfaces/Customer.md (hunk 7 above)
Comment (bc-sebastianszafraniec) [reply docs/interfaces/Customer.md:16]: I didn't know it is managed automatically thats why I removed `isStripeLinkAuthenticated`
In this situation I will not modify this file.

**Comment by bc-sebastianszafraniec** [comment]
I updated the Testing / Proof section in PR description

**Review by bc-peng — APPROVED** [review]
Congratulations on finishing a giant PR.

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 8]
```diff
@@ -235,27 +231,25 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
...
     }
 
     private async _executeWithoutRedirect(
         methodId: string,
         shouldSaveInstrument: boolean,
         shouldSetAsDefaultInstrument: boolean,
-    ): Promise<InternalCheckoutSelectors> {
-        const state = this._store.getState();
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+    ): Promise<PaymentIntegrationSelectors> {
```
Comment (PavlenkoM) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:250]: should returned value be as `PaymentIntegrationSelectors` maybe better to change to void? because we don't use this returned value after call
```suggestion
    ): Promise<void> {
```

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 9]
```diff
@@ -301,10 +293,10 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
         methodId: string,
         token: string,
         shouldSetAsDefaultInstrument: boolean,
-    ): Promise<InternalCheckoutSelectors> {
-        const state = this._store.getState();
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
-        const cartId = state.cart.getCart()?.id;
+    ): Promise<PaymentIntegrationSelectors | void> {
```
Comment (PavlenkoM) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:296]: the same as above
```suggestion
    ): Promise<void> {
```

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 10]
```diff
@@ -338,12 +328,15 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
...
-                params: { method: methodId },
-            }),
-        );
-        const paymentMethod = state.paymentMethods.getPaymentMethodOrThrow(methodId);
+        const state = await this.paymentIntegrationService.loadPaymentMethod(gatewayId, {
+            params: { method: methodId },
+        });
+        const paymentMethod = state.getPaymentMethodOrThrow(methodId);
+
+        if (!paymentMethod || !isStripeUPEPaymentMethodLike(paymentMethod)) {
```
Comment (PavlenkoM) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:336]: You are checking if payment method is existing inside type guard, I don't think that we need additional check there
```suggestion
        if (!isStripeUPEPaymentMethodLike(paymentMethod)) {
```

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts [hunk 11]
```diff
@@ -422,13 +417,15 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
         });
     }
 
+    // TODO: complexity of _processAdditionalAction method
+    // eslint-disable-next-line complexity
```
Comment (serhii-tkachenko) [inline packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:421]: Please, create a task to fix complexity issue of this method in the future

File: packages/stripe-integration/src/stripe-upe/create-stripe-upe-customer-strategy.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+const createStripeUPECustomerStrategy: CustomerStrategyFactory<StripeUPECustomerStrategy> = (
+    paymentIntegrationService,
+) => {
+    return new StripeUPECustomerStrategy(
+        paymentIntegrationService,
+        new StripeUPEScriptLoader(getScriptLoader()),
+    );
+};
+
+export default toResolvableModule(createStripeUPECustomerStrategy, [{ id: 'stripeupe' }]);
```
Comment (serhii-tkachenko) [inline packages/stripe-integration/src/stripe-upe/create-stripe-upe-customer-strategy.ts:20]: I see difference between customer and payment strategy. Should `stripeupe` be passed as `id` or as `gateway`?

https://github.com/bigcommerce/checkout-sdk-js/pull/2193/files#diff-e82d9aba9eaef2d31adf89f728c17e3ef625590948a6365d801f763867e25857R20

File: packages/payment-integrations-test-utils/src/test-utils/customer.mock.ts [hunk 13]
```diff
@@ -25,3 +25,17 @@ export function getCustomer(): Customer {
         },
     };
 }
+
+export function getGuestCustomer(): InternalCustomer {
```
[the remark this answers is not in the document]
Comment (bc-sebastianszafraniec) [reply packages/payment-integrations-test-utils/src/test-utils/customer.mock.ts:29]: This function was copied from https://github.com/bigcommerce/checkout-sdk-js/blob/195694ad468a70ab436c03b93a33f79057227be1/packages/core/src/customer/internal-customers.mock.ts#L7 and `InternalCustomer` type was already used there.

When I change type to `Customer` then there is type error. `Customer` is missing `customerId` property.

File: packages/payment-integration-api/src/payment/payment.ts [hunk 14]
```diff
@@ -292,7 +292,8 @@ interface StripeV3Intent {
     confirm: boolean;
 }
 
-interface StripeUPEIntent {
+export interface StripeUPEIntent {
+    cart_id: string;
```
[the remark this answers is not in the document]
Comment (bc-sebastianszafraniec) [reply packages/payment-integration-api/src/payment/payment.ts:296]: StripeUPEIntent interface was coppied from https://github.com/bigcommerce/checkout-sdk-js/blob/195694ad468a70ab436c03b93a33f79057227be1/packages/core/src/payment/payment.ts#L221. When `card_id` was added there it was added as required, so I think we can assume that it is required.

File: packages/stripe-integration/src/stripe-upe/stripe-upe.ts [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+        colorDanger?: string;
+        colorTextSecondary?: string;
+        colorTextPlaceholder?: string;
+        colorIcon?: string;
+        colorIconCardError?: string;
+        colorIconRedirect?: string;
+        spacingUnit?: string;
+        borderRadius?: string;
+    };
+
```
[the remark this answers is not in the document]
Comment (bc-sebastianszafraniec) [reply packages/stripe-integration/src/stripe-upe/stripe-upe.ts:301]: You mean empty line 301?

File: packages/stripe-integration/src/stripe-upe/stripe-upe-script-loader.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    getElements(stripeClient: StripeUPEClient, options: StripeElementsOptions): StripeElements {
+        let stripeElements = this.stripeWindow.bcStripeElements;
+
+        if (!stripeElements) {
+            stripeElements = stripeClient.elements(options);
+
+            Object.assign(this.stripeWindow, { bcStripeElements: stripeElements });
+        } else {
+            stripeElements.fetchUpdates();
```
[the remark this answers is not in the document]
Comment (bc-sebastianszafraniec) [reply packages/stripe-integration/src/stripe-upe/stripe-upe-script-loader.ts:55]: fetch_updates() function is provided by Stripe and it is communicating with Stripe. https://stripe.com/docs/js/elements_object/fetch_updates

File: packages/stripe-integration/src/stripe-upe/create-stripe-upe-customer-strategy.ts (hunk 12 above)
Comment (bc-sebastianszafraniec) [reply packages/stripe-integration/src/stripe-upe/create-stripe-upe-customer-strategy.ts:20]: I am not sure, but the way it is now passed (cusomer strategy - id, payment strategy - gateway), is the way it is working. When I set to `id` or `gateway` in both create strategy files, then it does't work.

File: packages/stripe-integration/src/stripe-upe/stripe-upe.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (serhii-tkachenko) [reply packages/stripe-integration/src/stripe-upe/stripe-upe.ts:289]: Yes

File: packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts (hunk 11 above)
Comment (serhii-tkachenko) [reply packages/stripe-integration/src/stripe-upe/stripe-upe-payment-strategy.ts:421]: Its better ask Maria about that

**Review by serhii-tkachenko — APPROVED** [review]
LGTM! Please squash your commits into one before merge
